specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1080 @@
+> Note: At this point there's no deduplication of atomic executions in the same `SCA`. An atomic execution with the same `AtomicExecParams` can't be run twice in the same `SCA`. This is a limitation of the MVP implementation and it will be fixed in the next iteration by adding a nonce (see [issue](https://github.com/filecoin-project/eudico/issues/147).
+
+- __Off-chain execution__: Each party only holds part of the state required for the execution. In order for the parties in the execution to be able to execute locally, they need to request the state locked in the other subnet. 
+    - The `CID` of the input state is shared between the different parties in the execution during the initialization stage, and is leveraged by each party to request from the other subnets the locked input states involved in the execution. 
+    - On the reception of the `CID` from the involved parties, peers can perform a request to the content resolution protocol for the `CID` of the input state in their source subnets by sending a `PullLocked` message. 
+    - Once every input state is received for all the `CID`s of the input state involved in the execution, each party runs the execution off-chain to compute the output state. This execution is performed by creating a temporal view of the state of the contract and merging the locked state from the other subnets. With the state ready, the every message of the atomic execution are implicitly applied ([see code](https://github.com/filecoin-project/eudico/blob/bb52565105f7fe716463b1e09dad7492569089f5/chain/consensus/hierarchical/atomic/exec/exec.go#L32) for further details).
+    - The `OutputState` for the execution is then returned for its commitment in the common parent.
+
+- __Commit atomic execution in parent subnet__: As the parties involved perform the off-chain execution of `OutputState`, they commit it in the `SCA` of the parent subnet. 
+    - The commitment is performed by sending a message to the `SubmitAtomicExec` of the `SCA` in the common parent using `SubmitExecParams` as an input. In `SubmitExecParams` users need to include their `LockedState` for the `OutputState` after the execution, and the `CID` of the atomic execution they are submitting.
+    - The `SCA` checks that the `CID` of the `LockedState` matches the one submitted by the other parties, and accepts the submission. The execution will stay in an `ExecInitialized` state until all the parties submit the right `LockedState` as an `OutputState`. When this happens, the execution is marked as `ExecSuccess` and the `SCA` triggers a top-down message to the corresponding subnet to the `Merge` function of the `LockableActor` involved in the execution on every subnet. This will trigger the merge the of the `OutputState` and the unlock of the locked state.
```

Comment:
```suggestion
    - The `SCA` checks that the `CID` of the `LockedState` matches the one submitted by the other parties, and accepts the submission. The execution will stay in an `ExecInitialized` state until all the parties submit the right `LockedState` as an `OutputState`. When this happens, the execution is marked as `ExecSuccess` and the `SCA` triggers a top-down message to the corresponding subnet to the `Merge` function of the `LockableActor` involved in the execution on every subnet. This will trigger the merging of the `OutputState` and the unlocking of the locked state.
```